### PR TITLE
[MLIR][NVVM] Enable nvvm intrinsics import  

### PR DESCRIPTION
**Pull request by ivanradanov**

### CODE REVIEW

**Comment by ivanradanov** [comment]
@ftynse

**Comment by ftynse** [comment]
@gysit is more knowledgeable about this part of the code so I'll defer to him. One general note is that I would split out NVVM into separate files, the same as we do on the MLIR->LLVMIR conversion side, rather than shove everything together.

**Comment by gysit** [comment]
I would second @ftynse here that we should separate the import of NVVM intrinsics. The import is very similar to the export in the sense that you can implement an interface for your target that implements the actual conversion. There are some difficulties to solve there since the intrinsics don't have a dialect associated in LLVM IR. The interface therefore needs to communicate which intrinsics it supports (https://discourse.llvm.org/t/rfc-extensible-llvm-ir-import/67256). 

In principle, you want to build a counterpart to LLVMDialectLLVMIRImportInterface inside the mlir/lib/Target/LLVMIR/Dialect/NVVM folder. It needs to overwrite the relevant methods (I guess you do not care about metadata import). The `getSupportedIntrinsics` method returns the supported intrinsics and `convertIntrinsicImpl`. Most of the code in this methods can be autogenerated using tablegen. Finally you need to make sure to register the new translation interface.

I think you are the first one that tries this mechanism for intrinsics (at least upstream). However, things should work analogous to the LLVMIR import.

**Comment by ivanradanov** [comment]
@gysit Thank you for the explanation, how does it look now?

File: mlir/include/mlir/Target/LLVMIR/Dialect/NVVM/LLVMIRToNVVMTranslation.h
Comment (gysit) [inline mlir/include/mlir/Target/LLVMIR/Dialect/NVVM/LLVMIRToNVVMTranslation.h:14]:
```suggestion
#ifndef MLIR_TARGET_LLVMIR_DIALECT_NVVM_LLVMIRTONVVMTRANSLATION_H
#define MLIR_TARGET_LLVMIR_DIALECT_NVVM_LLVMIRTONVVMTRANSLATION_H
```

File: mlir/include/mlir/Target/LLVMIR/Dialect/NVVM/LLVMIRToNVVMTranslation.h
Comment (gysit) [inline mlir/include/mlir/Target/LLVMIR/Dialect/NVVM/LLVMIRToNVVMTranslation.h:31]:
```suggestion
#endif // MLIR_TARGET_LLVMIR_DIALECT_NVVM_LLVMIRTONVVMTRANSLATION_H
```

File: mlir/test/Target/LLVMIR/Import/intrinsic.ll
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/intrinsic.ll:868]: Can you move the tests into an nvvmir.ll file in the same folder to follow the same structure as in the export?

Can you also add a test for every intrinsic? I know it seems like a lot of copy paste but it still seems helpful to exercise the import of all of them in case things get refactored in LLVM land for example.

**Review by gysit — COMMENTED** [review]
Nice! The structure looks good to me.

We probably want to increase the test coverage a little bit though.

**Comment by gysit** [comment]
Let me add @Dinistro as another reviewer since I will be leaving for a vacation tonight.

**Comment by ivanradanov** [comment]
@gysit @Dinistro 

I added a test which includes all of the currently supported nvvm intrinsics for import.

To be automatically imported the nvvm ops must inherit from NVVM_IntrOp, and most of them currently inherit from NVVM_Op. I will be looking into whether the remaining can be made to inherit from IntrOp but I think that is best left for a separate PR?

File: mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp
Comment (Dinistro) [inline mlir/lib/Target/LLVMIR/ConvertFromLLVMIR.cpp:15]: This include doesn't seem necessary.

File: mlir/test/Target/LLVMIR/Import/nvvmir.ll
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/nvvmir.ll:71]:
```suggestion
; TODO: Support the intrinsics below once they derive from NVVM_IntrOp rather than from NVVM_Op.
```
nit: Can you mention the reason here (at least this is what I understood from your comment on the PR.

**Review by gysit — APPROVED** [review]
LGMT modulo nit.

Yes I think it makes sense to separate the support for the remaining intrinsics into a followup PR.

File: mlir/test/Target/LLVMIR/Import/nvvmir.ll
Comment (Dinistro) [inline mlir/test/Target/LLVMIR/Import/nvvmir.ll:208]:
```suggestion
declare noundef i32 @llvm.nvvm.read.ptx.sreg.tid.x()
```

I don't think that we require function attributes. If we do, you should specify them explicitly and not through a comment that is ignored.

**Review by Dinistro — APPROVED** [review]
LGTM % minor nits.

**Comment by gysit** [comment]
@ivanradanov do you need help with landing the pr? 

Also if you do not have commit access yet you may want ask for it soonish https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

**Comment by ivanradanov** [comment]
> @ivanradanov do you need help with landing the pr?
> 
> Also if you do not have commit access yet you may want ask for it soonish https://llvm.org/docs/DeveloperPolicy.html#obtaining-commit-access.

If you could do it for me that would be great.

(Thank you for the link, I asked for commit access but do not have it yet)

**Comment by gysit** [comment]
Ah I see, the commit right process was faster then me :)

**Pull request merged**
